libnetdata/popen/popen.c

```diff
@@ -78,36 +78,75 @@ static void myp_del(pid_t pid) {
 #define PIPE_READ 0
 #define PIPE_WRITE 1
 
-/* custom_popene flag definitions */
-#define FLAG_CREATE_PIPE    1 // Create a pipe like popen() when set, otherwise set stdout to /dev/null
-#define FLAG_CLOSE_FD       2 // Close all file descriptors other than STDIN_FILENO, STDOUT_FILENO, STDERR_FILENO
-
 /*
- * Returns -1 on failure, 0 on success. When FLAG_CREATE_PIPE is set, on success set the FILE *fp pointer.
+ * Returns -1 on failure, 0 on success. When POPEN_FLAG_CREATE_PIPE is set, on success set the FILE *fp pointer.
  */
-static inline int custom_popene(const char *command, volatile pid_t *pidptr, char **env, uint8_t flags, FILE **fpp) {
+int custom_popene_internal_dont_use_directly(volatile pid_t *pidptr, char **env, uint8_t flags, FILE **fpp, const char *command, ...) {
```

Comment:
Why do you duplicate `command` in macros? Can't you just initialize `spawn_argv[0]` with it?

Reply:
This is a feature that may be needed so some other cases. I don't know. So, I preferred to have it available in case it is needed in the future.

So, this thing works like this:

1. execute command `command`
2. pass the following `char *argv[]` to it.
3. `argv[0]` is supposed to be the name of the command executed, as it is understood to the program.

So, we could for example say:

> execute `fping.plugin` but call it `fping2.plugin` so that once `fping.plugin` is run like that, it will search for `/etc/netdata/fping2.conf`, not `/etc/netdata/fping.conf`.

Rare use case, but why to lose it?

Also, I am not sure of the implications of running `/bin/sh` with `argv[]` as `/bin/sh -c "command"` instead of `sh -c "command"` may be. The original was doing `sh -c "command"`. The new does `/bin/sh -c "command"`.